retry.go

```diff
@@ -47,6 +49,42 @@ func X(x int, maxBackoff time.Duration, f func() bool) {
 	}
 }
 
+// XWithContext runs function f until f returns nil or the
+// number of retries exceeds x. Never more than x calls of f
+// are done. Calls to f have a sleep duration between them.
+// XWithContext will return a wrapped error around f's errors
+// if all attempts fail.
+// The attempts can be cancelled with ctx. If f does not cancel
+// when ctx is done, then the currently-running f will be allowed
+// to complete first.
+//
+// Example 1:
+//    var err error
+//    retry.XWithContext(context.Background(), 3, 5*time.Second, func(ctx context.Context) error {
+//        err = DoSomething()
+//        return err
+//    })
+func XWithContext(ctx context.Context, x int, maxBackoff time.Duration, f func(ctx context.Context) error) error {
+	var latestErr error
+	for i := 0; i < x; i++ {
+		timer := time.NewTimer(backoff(i, maxBackoff))
```

Comment:
Can we move this out of the for loop and just use reset?

Reply:
Good call, Quickly creating Timers can cause lots of heap pressure.

Reply:
Moved it out. It gets a little funky when you reset twice, so there's still a conditional in the loop for reset/new.